README.md
[![Serverless Nuxtjs Tencent Cloud](https://img.serverlesscloud.cn/2020310/1583829094342-nuxt.js%20_%E9%95%BF.png)](http://serverless.com)

# 腾讯云 Nuxt.js Serverless Component

## 简介

**腾讯云[Nuxt.js](https://github.com/nuxt/nuxt.js)组件** - 通过使用[**Tencent Serverless Framework**](https://github.com/serverless/components/tree/cloud) , 基于云上 Serverless 服务（如 API 网关、云函数等），实现“0”配置，便捷开发，极速部署采用 Nuxt.js 框架的网页应用，Nuxt.js 组件支持丰富的配置扩展，提供了目前便捷实用，开发成本低的网页应用项目的开发/托管能力。

特性介绍：

- [x] **按需付费** - 按照请求的使用量进行收费，没有请求时无需付费
- [x] **"0"配置** - 只需要关心项目代码，之后部署即可，Serverless Framework 会搞定所有配置。
- [x] **极速部署** - 部署速度快，仅需几秒，部署你的整个应用。
- [x] **实时日志** - 通过实时日志的输出查看业务状态，便于直接在云端开发应用。
- [x] **云端调试** - 可在云端直接进行项目调试，从而避免本地环境的差异。
- [x] **便捷协作** - 通过云端控制台的状态信息和部署日志，方便进行多人协作开发。

## 快速开始

0. [**准备**](#0-准备)
1. [**安装**](#1-安装)
1. [**配置**](#2-配置)
1. [**部署**](#3-部署)
1. [**开发调试**](#4-开发调试)
1. [**查看状态**](#5-查看部署状态)
1. [**移除**](#6-移除)

更多资源：

- [**账号配置**](#账号配置)
- [**更多组件**](#更多组件)
- [**FAQ**](#FAQ)

### 1. 安装

通过 npm 全局安装 [serverless cli](https://github.com/serverless/serverless)

```bash
$ npm install -g serverless
```

### 2. 创建

通过如下命令和模板链接，快速创建一个 nuxtjs 应用：

```bash
$ serverless init nuxtjs-starter --name example
$ cd example
```

### 3. 部署

在 `serverless.yml` 文件所在的项目根目录，运行以下指令，将会弹出二维码，直接扫码授权进行部署：

```
serverless deploy
```

> **说明**：如果鉴权失败，请参考 [权限配置](https://cloud.tencent.com/document/product/1154/43006) 进行授权。

### 4. 配置

nuxtjs 组件支持 0 配置部署，也就是可以直接通过配置文件中的默认值进行部署。但你依然可以修改更多可选配置来进一步开发该 nuxtjs 项目。

以下是 nuxtjs 组件的 `serverless.yml`配置示例：

```yml
# serverless.yml
component: nuxtjs # (必填) 组件名称，此处为nuxtjs
name: nuxtjsDemo # (必填) 实例名称
org: orgDemo # (可选) 用于记录组织信息，默认值为您的腾讯云账户 appid
app: appDemo # (可选) 该 nuxt.js 项目名称
stage: dev # (可选) 用于区分环境信息，默认值是 dev

inputs:
  src:
    src: ./
    exclude:
      - .env
  region: ap-guangzhou
  runtime: Nodejs10.15
  apigatewayConf:
    protocols:
      - http
      - https
    environment: release
```

- 点此查看[更多配置及说明](https://github.com/serverless-components/tencent-nuxtjs/tree/master/docs/configure.md)

### 4. 开发调试

部署了 Nuxt.js 应用后，可以通过开发调试能力对该项目进行二次开发，从而开发一个生产应用。在本地修改和更新代码后，不需要每次都运行 `serverless deploy` 命令来反复部署。你可以直接通过 `serverless dev` 命令对本地代码的改动进行检测和自动上传。

可以通过在 `serverless.yml`文件所在的目录下运行 `serverless dev` 命令开启开发调试能力。

`serverless dev` 同时支持实时输出云端日志，每次部署完毕后，对项目进行访问，即可在命令行中实时输出调用日志，便于查看业务情况和排障。

### 5. 查看部署状态

在`serverless.yml`文件所在的目录下，通过如下命令查看部署状态：

```
$ serverless info
```

### 6. 移除

在`serverless.yml`文件所在的目录下，通过以下命令移除部署通过以下命令移除部署的 API 网关，移除后该组件会对应删除云上部署时所创建的所有相关资源。

```bash
$ serverless remove
```

和部署类似，支持通过 `serverless remove --debug` 命令查看移除过程中的实时日志信息，`serverless`是 `serverless` 命令的缩写。

### 账号配置

当前默认支持 CLI 扫描二维码登录，如您希望配置持久的环境变量/秘钥信息，也可以本地创建 `.env` 文件

```bash
$ touch .env # 腾讯云的配置信息
```

在 `.env` 文件中配置腾讯云的 SecretId 和 SecretKey 信息并保存

如果没有腾讯云账号，可以在此 [注册新账号](https://cloud.tencent.com/register)。

如果已有腾讯云账号，可以在 [API 密钥管理](https://console.cloud.tencent.com/cam/capi) 中获取 `SecretId` 和`SecretKey`.

```text
# .env
TENCENT_SECRET_ID=123
TENCENT_SECRET_KEY=123
```

> 注意：海外 ip 登录时，需要在`.env`文件中添加`SERVERLESS_PLATFORM_VENDOR=tencent` ，使 serverless 默认使用 tencent 组件

## 更多组件

可以在 [Serverless Components](https://github.com/serverless/components) repo 中查询更多组件的信息。

## 项目迁移

如果项目使用了自定义 Node.js 服务，比如 express 或者 koa，你需要做如下改造工作。

### 自定义 express 服务

如果你的 Nuxt.js 项目本身运行就是基于 `express` 自定义服务的，那么你需要在项目中自定义入口文件 `sls.js`，需要参考你的服务启动文件进行修改，以下是一个模板文件：

```js
const express = require('express')
const { loadNuxt } = require('nuxt')

async function createServer() {
  // not report route for custom monitor
  const noReportRoutes = ['/_nuxt', '/static', '/favicon.ico']

  const server = express()
  const nuxt = await loadNuxt('start')

  server.all('*', (req, res, next) => {
    noReportRoutes.forEach((route) => {
      if (req.path.indexOf(route) === 0) {
        req.__SLS_NO_REPORT__ = true
      }
    })
    return nuxt.render(req, res, next)
  })

  // define binary type for response
  // if includes, will return base64 encoded, very useful for images
  server.binaryTypes = ['*/*']

  return server
}

module.exports = createServer
```

### 自定义 koa 服务

如果你的项目使用的是 Koa 作为 Node.js 服务，需要在项目中自定义入口文件 `sls.js`，需要参考你的服务启动文件进行修改，以下是一个模板文件：

```js
const Koa = require('koa')
const { loadNuxt } = require('nuxt')

async function createServer() {
  const server = new Koa()
  const nuxt = await loadNuxt('start')

  server.use((ctx) => {
    ctx.status = 200
    ctx.respond = false
    ctx.req.ctx = ctx

    nuxt.render(ctx.req, ctx.res)
  })

  // define binary type for response
  // if includes, will return base64 encoded, very useful for images
  server.binaryTypes = ['*/*']

  return server
}

module.exports = createServer
```

## 自定义监控

> **目前仅支持自定义 `Express` 服务的项目**

当在部署 Nuxt.js 应用时，如果 `serverless.yml` 中未指定 `role`，默认会尝试绑定 `QCS_SCFExcuteRole`，并且开启自定义监控，帮助用户收集应用监控指标。对于为自定义入口文件的项目，会默认上报除含有 `/_nuxt`、`/static` 和 `/favicon.ico` 的路由。如果你想自定义上报自己的路由性能，那么可以自定义 `sls.js` 入口文件，对于无需上报的路由，在 express 服务的 `req` 对象上添加 `__SLS_NO_REPORT__` 属性值为 `true` 即可。比如：

```js
server.get('/no-report', (req, res, next) => {
  req.__SLS_NO_REPORT__ = true
  return nuxt.render(req, res, next)
})
```

那么用户在访问 `GET /no-report` 路由时，就不会上报自定义监控指标。

## 文件上传

[文件上传教程](https://github.com/serverless-components/tencent-nuxtjs/tree/master/docs/upload.md)

## License

MIT License

Copyright (c) 2020 Tencent Cloud, Inc.


CHANGELOG.md
## [0.2.1](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.2.0...v0.2.1) (2021-01-26)


### Bug Fixes

* change cos access by policy ([7884f8d](https://github.com/serverless-components/tencent-nuxtjs/commit/7884f8de205951956861bfcda6e1a892e1b18018))

# [0.2.0](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.1.8...v0.2.0) (2021-01-25)


### Features

* support apigw base64 encode ([ef3d53c](https://github.com/serverless-components/tencent-nuxtjs/commit/ef3d53c19d61dc0a952290f3c2cb84c767f8135e))

## [0.1.8](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.1.7...v0.1.8) (2020-12-21)


### Bug Fixes

* update remove flow ([#17](https://github.com/serverless-components/tencent-nuxtjs/issues/17)) ([96879d4](https://github.com/serverless-components/tencent-nuxtjs/commit/96879d4d541a5598cafc1acb91dcacf0a62871ed))

## [0.1.7](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.1.6...v0.1.7) (2020-12-15)


### Bug Fixes

* update to deployment to serial flow ([850dafe](https://github.com/serverless-components/tencent-nuxtjs/commit/850dafe6dcb89bdc9fdcc85affff27c62b05e90b))

## [0.1.6](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.1.5...v0.1.6) (2020-10-20)


### Bug Fixes

* support replace deployment ([31b29c7](https://github.com/serverless-components/tencent-nuxtjs/commit/31b29c7dd0b72e8517736810e775b8ed1847dc90))

## [0.1.5](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.1.4...v0.1.5) (2020-10-12)


### Bug Fixes

* support all parameters for apigw ([15805a6](https://github.com/serverless-components/tencent-nuxtjs/commit/15805a6b74751407e217ee5bb80dc3e127f92842))

## [0.1.4](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.1.3...v0.1.4) (2020-10-10)


### Bug Fixes

* support customize entry file ([50ace57](https://github.com/serverless-components/tencent-nuxtjs/commit/50ace57900ebf9b4f405ec38204eb85a78e7aa77))

## [0.1.3](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.1.2...v0.1.3) (2020-09-25)


### Bug Fixes

* support koa server ([#9](https://github.com/serverless-components/tencent-nuxtjs/issues/9)) ([70d1564](https://github.com/serverless-components/tencent-nuxtjs/commit/70d15643902dcca3111501c12fe7837525a13772))

## [0.1.1](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.1.0...v0.1.1) (2020-09-03)


### Bug Fixes

* update deps ([1a883c5](https://github.com/serverless-components/tencent-nuxtjs/commit/1a883c56fe980aaa7bcc34043cabb99dc0f6cec8))

# [0.1.0](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.0.14...v0.1.0) (2020-09-02)


### Features

* add mark for api request ([db0328f](https://github.com/serverless-components/tencent-nuxtjs/commit/db0328f498cd45d16e5bc1d6d17cdc5b5c860c42))

## [0.0.14](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.0.13...v0.0.14) (2020-08-27)


### Bug Fixes

* optimize deploy flow ([2768d16](https://github.com/serverless-components/tencent-nuxtjs/commit/2768d16b50aeea7343dae468421cb7eaf9902f1c))

## [0.0.13](https://github.com/serverless-components/tencent-nuxtjs/compare/v0.0.12...v0.0.13) (2020-08-26)


### Bug Fixes

* apigw custom domain update bug ([8c1225b](https://github.com/serverless-components/tencent-nuxtjs/commit/8c1225bf6ef306a96abfd05a522eb3e0ea17dcc3))
* cache http server ([8cec64e](https://github.com/serverless-components/tencent-nuxtjs/commit/8cec64eaf62256ad64c48348dd653d54122a85ce))
* optimize static cos output ([5d45e22](https://github.com/serverless-components/tencent-nuxtjs/commit/5d45e22b2290628e2cc085c29a1c225c05ea6f23))
* read bucket and object from srcOriginal ([e5c000d](https://github.com/serverless-components/tencent-nuxtjs/commit/e5c000da0d31c7f9be9038b0605763b26c75fd77))
* static cos bucket has appid suffix ([9ff78a9](https://github.com/serverless-components/tencent-nuxtjs/commit/9ff78a98d87f574e38bcd8d1fa393d83ae7bc0be))
* support eip config ([e5034af](https://github.com/serverless-components/tencent-nuxtjs/commit/e5034af0b6ebb1ce929c7fe6cc37597d7ef8096b))
* traffic zero display bug ([f95f479](https://github.com/serverless-components/tencent-nuxtjs/commit/f95f479662f1ce385832d30031f4b0981e6d4958))
* typo ([584aa55](https://github.com/serverless-components/tencent-nuxtjs/commit/584aa55bff4b929275dff011b921d4c4d7437307))
* update deps ([af359f7](https://github.com/serverless-components/tencent-nuxtjs/commit/af359f735c42d866f96ef13bb6fc68b29f05c021))
* upgrade deps ([b0dd2b3](https://github.com/serverless-components/tencent-nuxtjs/commit/b0dd2b34d56049266fab4862aaee5597a7ef66e6))
* upgrade deps ([f0ebc4d](https://github.com/serverless-components/tencent-nuxtjs/commit/f0ebc4dd747314aaadff0e5c799460e844971b99))
* upgrade deps ([1c4d4f8](https://github.com/serverless-components/tencent-nuxtjs/commit/1c4d4f83d829ff8c031e467dc22a4457c7a4e698))
* upgrade tencent-component-toolkit for deleting compatibility ([e5e3382](https://github.com/serverless-components/tencent-nuxtjs/commit/e5e3382e510999a1832f5888765808f213548c28))
* upgrade tencent-serverless-http ([f06f3ca](https://github.com/serverless-components/tencent-nuxtjs/commit/f06f3caaf62d9d3619676cb4fb99b1050f1df6b1))


### Features

* add layer support ([c2421ec](https://github.com/serverless-components/tencent-nuxtjs/commit/c2421ecb5b0dca91698638a124dd79668c10ecdc))
* init v2 ([a11e63b](https://github.com/serverless-components/tencent-nuxtjs/commit/a11e63b73301175aa9b4c7e36152fbededb8c78b))
* optimize zip flow ([5e3d2e1](https://github.com/serverless-components/tencent-nuxtjs/commit/5e3d2e1cae7525486274a71676f79afbc6cb6267))
* support api gw metrics ([1a14b24](https://github.com/serverless-components/tencent-nuxtjs/commit/1a14b2447b3cbabe57b238bf6adf6b37ca8d0fbf))
* support disable creating apigw ([87e2499](https://github.com/serverless-components/tencent-nuxtjs/commit/87e24998a9e98e1da1419696141344e2b35838f9))

src/serverless.js
const { Component } = require('@serverless/core')
const { Scf, Apigw, Cns, Cam, Metrics, Cos, Cdn, Domain } = require('tencent-component-toolkit')
const { TypeError } = require('tencent-component-toolkit/src/utils/error')
const {
  deepClone,
  uploadCodeToCos,
  getDefaultProtocol,
  prepareInputs,
  prepareStaticCosInputs,
  prepareStaticCdnInputs
} = require('./utils')
const CONFIGS = require('./config')

class ServerlessComopnent extends Component {
  getCredentials() {
    const { tmpSecrets } = this.credentials.tencent

    if (!tmpSecrets || !tmpSecrets.TmpSecretId) {
      throw new TypeError(
        'CREDENTIAL',
        'Cannot get secretId/Key, your account could be sub-account and does not have the access to use SLS_QcsRole, please make sure the role exists first, then visit https://cloud.tencent.com/document/product/1154/43006, follow the instructions to bind the role to your account.'
      )
    }

    return {
      SecretId: tmpSecrets.TmpSecretId,
      SecretKey: tmpSecrets.TmpSecretKey,
      Token: tmpSecrets.Token
    }
  }

  getAppId() {
    return this.credentials.tencent.tmpSecrets.appId
  }

  async deployFunction(credentials, inputs, regionList) {
    if (!inputs.role) {
      try {
        const camClient = new Cam(credentials)
        const roleExist = await camClient.CheckSCFExcuteRole()
        if (roleExist) {
          inputs.role = 'QCS_SCFExcuteRole'
        }
      } catch (e) {
        // no op
      }
    }

    const outputs = {}
    const appId = this.getAppId()

    const funcDeployer = async (curRegion) => {
      const code = await uploadCodeToCos(this, appId, credentials, inputs, curRegion)
      const scf = new Scf(credentials, curRegion)
      const tempInputs = {
        ...inputs,
        code
      }
      const scfOutput = await scf.deploy(deepClone(tempInputs))
      outputs[curRegion] = {
        functionName: scfOutput.FunctionName,
        runtime: scfOutput.Runtime,
        namespace: scfOutput.Namespace
      }

      this.state[curRegion] = {
        ...(this.state[curRegion] ? this.state[curRegion] : {}),
        ...outputs[curRegion]
      }

      // default version is $LATEST
      outputs[curRegion].lastVersion = scfOutput.LastVersion
        ? scfOutput.LastVersion
        : this.state.lastVersion || '$LATEST'

      // default traffic is 1.0, it can also be 0, so we should compare to undefined
      outputs[curRegion].traffic =
        scfOutput.Traffic !== undefined
          ? scfOutput.Traffic
          : this.state.traffic !== undefined
          ? this.state.traffic
          : 1

      if (outputs[curRegion].traffic !== 1 && scfOutput.ConfigTrafficVersion) {
        outputs[curRegion].configTrafficVersion = scfOutput.ConfigTrafficVersion
        this.state.configTrafficVersion = scfOutput.ConfigTrafficVersion
      }

      this.state.lastVersion = outputs[curRegion].lastVersion
      this.state.traffic = outputs[curRegion].traffic
    }

    for (let i = 0; i < regionList.length; i++) {
      const curRegion = regionList[i]
      await funcDeployer(curRegion)
    }
    this.save()
    return outputs
  }

  // try to add dns record
  async tryToAddDnsRecord(credentials, customDomains) {
    try {
      const cns = new Cns(credentials)
      for (let i = 0; i < customDomains.length; i++) {
        const item = customDomains[i]
        if (item.domainPrefix) {
          await cns.deploy({
            domain: item.subDomain.replace(`${item.domainPrefix}.`, ''),
            records: [
              {
                subDomain: item.domainPrefix,
                recordType: 'CNAME',
                recordLine: '默认',
                value: item.cname,
                ttl: 600,
                mx: 10,
                status: 'enable'
              }
            ]
          })
        }
      }
    } catch (e) {
      console.log('METHOD_tryToAddDnsRecord', e.message)
    }
  }

  async addDnsRecord(credentials, targetDomain, targetValue) {
    try {
      const cns = new Cns(credentials)
      const domain = new Domain(credentials)
      const domainInfo = await domain.check(targetDomain)
      console.log(`Domain check result: ${JSON.stringify(domainInfo)}`)
      if (domainInfo) {
        // prefix is null will set main domain(set @ as prefix)
        const domainPrefix = domainInfo.subDomain || '@'
        console.log('cns deploy, prefix:' + domainPrefix)
        const cnsOutput = await cns.deploy({
          records: [
            {
              domain: domainInfo.domain,
              subDomain: domainPrefix,
              recordType: 'CNAME',
              recordLine: '默认',
              value: targetValue,
              ttl: 600,
              mx: 10,
              status: 'enable'
            }
          ]
        })
        console.log(`cns deploy result: ${JSON.stringify(cnsOutput)}`)
      }
    } catch (e) {
      console.log('METHOD_tryToAddDnsRecord', e.message)
    }
  }

  // try to add dns record for custom domains
  async tryToAddDnsRecordForCusDomain(credentials, customDomains) {
    for (let i = 0; i < customDomains.length; i++) {
      const item = customDomains[0]
      await this.addDnsRecord(credentials, item.subDomain, item.cname)
    }
  }

  // try to add dns record for cdn
  async tryToAddDnsRecordForCdn(credentials, cdnInfo) {
    await this.addDnsRecord(credentials, cdnInfo.domain, cdnInfo.cname)
  }

  async deployApigateway(credentials, inputs, regionList) {
    if (inputs.isDisabled) {
      return {}
    }

    const getServiceId = (instance, region) => {
      const regionState = instance.state[region]
      return inputs.serviceId || (regionState && regionState.serviceId)
    }

    const deployTasks = []
    const outputs = {}
    regionList.forEach((curRegion) => {
      const apigwDeployer = async () => {
        const apigw = new Apigw(credentials, curRegion)

        const oldState = this.state[curRegion] || {}
        const apigwInputs = {
          ...inputs,
          oldState: {
            apiList: oldState.apiList || [],
            customDomains: oldState.customDomains || []
          }
        }
        // different region deployment has different service id
        apigwInputs.serviceId = getServiceId(this, curRegion)
        const apigwOutput = await apigw.deploy(deepClone(apigwInputs))
        outputs[curRegion] = {
          serviceId: apigwOutput.serviceId,
          subDomain: apigwOutput.subDomain,
          environment: apigwOutput.environment,
          url: `${getDefaultProtocol(inputs.protocols)}://${apigwOutput.subDomain}/${
            apigwOutput.environment
          }/`
        }

        if (apigwOutput.customDomains) {
          if (apigwInputs.autoCreateDns === true) {
            console.log('Starting try add dns record for customDomains')
            await this.tryToAddDnsRecordForCusDomain(credentials, apigwOutput.customDomains)
          }
          outputs[curRegion].customDomains = apigwOutput.customDomains
        }
        this.state[curRegion] = {
          created: true,
          ...(this.state[curRegion] ? this.state[curRegion] : {}),
          ...outputs[curRegion],
          apiList: apigwOutput.apiList
        }
      }
      deployTasks.push(apigwDeployer())
    })

    await Promise.all(deployTasks)

    this.save()
    return outputs
  }

  // deploy static to cos, and setup cdn
  async deployStatic(credentials, inputs, region) {
    const { zipPath } = this.state
    const appId = this.getAppId()
    const deployStaticOutpus = {}

    if (zipPath) {
      console.log(`Deploy static for ${CONFIGS.compFullname} application`)
      // 1. deploy to cos
      const staticCosInputs = await prepareStaticCosInputs(this, inputs, appId, zipPath)

      const cos = new Cos(credentials, region)
      const cosOutput = {
        region
      }
      for (let i = 0; i < staticCosInputs.length; i++) {
        const curInputs = staticCosInputs[i]
        console.log(`Starting deploy directory ${curInputs.src} to cos bucket ${curInputs.bucket}`)
        const deployRes = await cos.deploy(curInputs)
        cosOutput.cosOrigin = `${curInputs.bucket}.cos.${region}.myqcloud.com`
        cosOutput.bucket = deployRes.bucket
        cosOutput.url = `https://${curInputs.bucket}.cos.${region}.myqcloud.com`
        console.log(`Deploy directory ${curInputs.src} to cos bucket ${curInputs.bucket} success`)
      }
      deployStaticOutpus.cos = cosOutput

      // 2. deploy cdn
      if (inputs.cdnConf) {
        const cdn = new Cdn(credentials)
        const cdnInputs = await prepareStaticCdnInputs(this, inputs, cosOutput.cosOrigin)
        console.log(`Starting deploy cdn ${cdnInputs.domain}`)
        const cdnDeployRes = await cdn.deploy(cdnInputs)
        const protocol = cdnInputs.https ? 'https' : 'http'
        const cdnOutput = {
          domain: cdnDeployRes.domain,
          url: `${protocol}://${cdnDeployRes.domain}`,
          cname: cdnDeployRes.cname
        }
        deployStaticOutpus.cdn = cdnOutput

        if (cdnInputs.autoCreateDns) {
          console.log('Starting try add dns record for cdn')
          await this.tryToAddDnsRecordForCdn(credentials, cdnOutput)
        }

        console.log(`Deploy cdn ${cdnInputs.domain} success`)
      }

      console.log(`Deployed static for ${CONFIGS.compFullname} application successfully`)

      return deployStaticOutpus
    }

    return null
  }

  async deploy(inputs) {
    console.log(`Deploying ${CONFIGS.compFullname} App...`)

    const credentials = this.getCredentials()

    // 对Inputs内容进行标准化
    const { regionList, functionConf, apigatewayConf } = await prepareInputs(
      this,
      credentials,
      inputs
    )

    // 部署函数 + API网关
    const outputs = {}
    if (!functionConf.code.src) {
      outputs.templateUrl = CONFIGS.templateUrl
    }

    const deployTasks = [this.deployFunction(credentials, functionConf, regionList, outputs)]
    // support apigatewayConf.isDisabled
    if (apigatewayConf.isDisabled !== true) {
      deployTasks.push(this.deployApigateway(credentials, apigatewayConf, regionList, outputs))
    } else {
      this.state.apigwDisabled = true
    }
    const [functionOutputs, apigwOutputs = {}] = await Promise.all(deployTasks)

    // optimize outputs for one region
    if (regionList.length === 1) {
      const [oneRegion] = regionList
      outputs.region = oneRegion
      outputs['apigw'] = apigwOutputs[oneRegion]
      outputs['scf'] = functionOutputs[oneRegion]
    } else {
      outputs['apigw'] = apigwOutputs
      outputs['scf'] = functionOutputs
    }

    // start deploy static cdn
    if (inputs.staticConf) {
      const staticOutputs = {}
      const staticTasks = []
      regionList.forEach((region) => {
        const staticDeployer = async () => {
          const res = await this.deployStatic(credentials, inputs.staticConf, region)
          staticOutputs[region] = res
        }
        staticTasks.push(staticDeployer())
      })

      await Promise.all(staticTasks)

      if (regionList.length === 1) {
        const [oneRegion] = regionList
        this.state.staticConf = staticOutputs[oneRegion]
        outputs.staticConf = staticOutputs[oneRegion]
      } else {
        this.state.staticConf = staticOutputs
        outputs.staticConf = staticOutputs
      }
    }

    this.state.region = regionList[0]
    this.state.regionList = regionList

    this.state.lambdaArn = functionConf.name

    return outputs
  }

  async removeStatic() {
    // remove static
    const { region, staticConf } = this.state
    if (staticConf) {
      console.log(`Removing static config`)
      const credentials = this.getCredentials()
      // 1. remove cos
      if (staticConf.cos) {
        const cos = new Cos(credentials, region)
        await cos.remove(staticConf.cos)
      }
      // 2. remove cdn
      if (staticConf.cdn) {
        const cdn = new Cdn(credentials)
        try {
          await cdn.remove(staticConf.cdn)
        } catch (e) {
          // no op
        }
      }
      console.log(`Remove static config success`)
    }
  }

  async remove() {
    console.log(`Removing ${CONFIGS.compFullname} App...`)

    const { state } = this
    const { regionList = [] } = state

    const credentials = this.getCredentials()

    const removeHandlers = []
    for (let i = 0; i < regionList.length; i++) {
      const curRegion = regionList[i]
      const curState = state[curRegion]
      const scf = new Scf(credentials, curRegion)
      const apigw = new Apigw(credentials, curRegion)
      const handler = async () => {
        await scf.remove({
          functionName: curState.functionName,
          namespace: curState.namespace
        })
        // if disable apigw, no need to remove
        if (state.apigwDisabled !== true) {
          await apigw.remove({
            created: curState.created,
            environment: curState.environment,
            serviceId: curState.serviceId,
            apiList: curState.apiList,
            customDomains: curState.customDomains
          })
        }
      }
      removeHandlers.push(handler())
    }

    await Promise.all(removeHandlers)

    // remove static
    await this.removeStatic()

    this.state = {}
  }

  async metrics(inputs = {}) {
    console.log(`Get ${CONFIGS.compFullname} Metrics Datas...`)
    if (!inputs.rangeStart || !inputs.rangeEnd) {
      throw new TypeError(
        `PARAMETER_${CONFIGS.compName.toUpperCase()}_METRICS`,
        'rangeStart and rangeEnd are require inputs'
      )
    }
    const { region } = this.state
    if (!region) {
      throw new TypeError(
        `PARAMETER_${CONFIGS.compName.toUpperCase()}_METRICS`,
        'No region property in state'
      )
    }
    const { functionName, namespace, functionVersion } = this.state[region] || {}
    if (functionName) {
      const options = {
        funcName: functionName,
        namespace: namespace,
        version: functionVersion,
        region,
        timezone: inputs.tz
      }
      const curState = this.state[region]
      if (curState.serviceId) {
        options.apigwServiceId = curState.serviceId
        options.apigwEnvironment = curState.environment || 'release'
      }
      const credentials = this.getCredentials()
      const mertics = new Metrics(credentials, options)
      const metricResults = await mertics.getDatas(
        inputs.rangeStart,
        inputs.rangeEnd,
        Metrics.Type.All
      )
      return metricResults
    }
    throw new TypeError(
      `PARAMETER_${CONFIGS.compName.toUpperCase()}_METRICS`,
      'Function name not define'
    )
  }
}

module.exports = ServerlessComopnent
